File: cards.java
import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.*;
import javafx.stage.Stage;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.*;

public class Cards extends Application {

    private static final String FILENAME = "data.json";

    public static void main(String[] args) {
        launch(args);
    }

    @Override
    public void start(Stage primaryStage) {
        primaryStage.setTitle("Card Data Entry");

        Label cardNameLabel = new Label("Card Name:");
        TextField cardNameField = new TextField();

        Label titleLabel = new Label("Title:");
        TextField titleField = new TextField();

        Label textLabel = new Label("Text:");
        TextArea textField = new TextArea();

        Label imageLabel = new Label("Image Address:");
        TextField imageField = new TextField();

        Label tagsLabel = new Label("Tags (comma-separated):");
        TextField tagsField = new TextField();

        Button saveButton = new Button("Save to JSON");
        saveButton.setOnAction(e -> {
            String cardName = cardNameField.getText();
            String title = titleField.getText();
            String text = textField.getText();
            String image = imageField.getText();
            String[] tags = tagsField.getText().split(",");

            Gson gson = new Gson();
            JsonObject cardData = new JsonObject();
            cardData.addProperty("title", title);
            cardData.addProperty("text", text);
            cardData.addProperty("image", image);
            cardData.add("tags", gson.toJsonTree(tags));

            File file = new File(FILENAME);

            try {
                JsonParser parser = new JsonParser();
                JsonObject existingData = new JsonObject();
                if (file.exists()) {
                    BufferedReader br = new BufferedReader(new FileReader(FILENAME));
                    existingData = parser.parse(br).getAsJsonObject();
                    br.close();
                }

                if (existingData.has(cardName)) {
                    Alert alert = new Alert(Alert.AlertType.ERROR);
                    alert.setTitle("Error");
                    alert.setHeaderText(null);
                    alert.setContentText("Card name already exists in the JSON file.");
                    alert.showAndWait();
                } else {
                    existingData.add(cardName, cardData);

                    FileWriter writer = new FileWriter(FILENAME);
                    gson.toJson(existingData, writer);
                    writer.close();

                    Alert alert = new Alert(Alert.AlertType.INFORMATION);
                    alert.setTitle("Success");
                    alert.setHeaderText(null);
                    alert.setContentText("Data appended to " + FILENAME);
                    alert.showAndWait();
                }
            } catch (IOException ioException) {
                ioException.printStackTrace();
            }
        });

        VBox layout = new VBox(10);
        layout.getChildren().addAll(
                cardNameLabel, cardNameField,
                titleLabel, titleField,
                textLabel, textField,
                imageLabel, imageField,
                tagsLabel, tagsField,
                saveButton
        );

        Scene scene = new Scene(layout, 400, 400);
        primaryStage.setScene(scene);
        primaryStage.show();
    }
}

File: archive/cards.py
import tkinter as tk
from tkinter import messagebox
import json
import os
import re

def save_to_json():
    card_name = card_name_entry.get()
    title = title_entry.get()
    text = text_entry.get("1.0", "end-1c")
    image = image_entry.get()
    tags = tags_entry.get().split(',')

    # Find and replace hyperlinks with HTML anchor tags
    link_pattern = r'(https?://\S+)'
    text_with_links = re.sub(link_pattern, r'<a href="\1">\1</a>', text)

    data = {
        "title": title,
        "text": text_with_links,
        "image": image,
        "tags": [tag.strip() for tag in tags]
    }

    output_file = "data.json"

    if os.path.exists(output_file):
        with open(output_file, 'r') as json_file:
            existing_data = json.load(json_file)
    else:
        existing_data = {}

    if card_name in existing_data:
        messagebox.showerror("Error", "Card name already exists in the JSON file.")
        return

    existing_data[card_name] = data

    with open(output_file, 'w') as json_file:
        json.dump(existing_data, json_file, indent=4)

    messagebox.showinfo("Success", "Data appended to " + output_file)

# Create the main window
root = tk.Tk()
root.title("Card Data Entry")

# Create labels and entry fields
card_name_label = tk.Label(root, text="Card Name:")
card_name_label.pack()
card_name_entry = tk.Entry(root)
card_name_entry.pack()

title_label = tk.Label(root, text="Title:")
title_label.pack()
title_entry = tk.Entry(root)
title_entry.pack()

text_label = tk.Label(root, text="Text:")
text_label.pack()
text_entry = tk.Text(root, height=5, width=40)
text_entry.pack()

image_label = tk.Label(root, text="Image Address:")
image_label.pack()
image_entry = tk.Entry(root)
image_entry.pack()

tags_label = tk.Label(root, text="Tags (comma-separated):")
tags_label.pack()
tags_entry = tk.Entry(root)
tags_entry.pack()

# Create a save button
save_button = tk.Button(root, text="Save to JSON", command=save_to_json)
save_button.pack()

# Start the GUI event loop
root.mainloop()
